File: ios/AlpineApp/AlpineApp/ViewModels/PeersViewModel.swift
import Foundation
import Observation

@Observable
final class PeersViewModel {
    var peers: [PeerDetail] = []
    var isLoading = false
    var message: String?

    private let settings: SettingsStore
    private let secureStorage: SecureStorage

    init(settings: SettingsStore, secureStorage: SecureStorage) {
        self.settings = settings
        self.secureStorage = secureStorage
    }

    private func createRpcService() -> AlpineRpcService? {
        let tlsConfig = TlsConfig(
            enabled: settings.tlsEnabled,
            mode: settings.tlsMode,
            certFingerprint: settings.tlsCertFingerprint
        )
        guard let url = tlsConfig.buildURL(host: settings.host, port: settings.port) else {
            return nil
        }
        let apiKey = secureStorage.read(key: "apiKey") ?? ""
        let client = JsonRpcClient(baseURL: url, apiKey: apiKey, tlsConfig: tlsConfig)
        return AlpineRpcService(client: client)
    }

    func loadPeers() async {
        guard let rpc = createRpcService() else {
            message = "Invalid server configuration"
            return
        }
        isLoading = true
        do {
            let peerIds = try await rpc.getAllPeers()
            var details: [PeerDetail] = []
            for id in peerIds {
                let detail = try await rpc.getPeer(peerId: id)
                details.append(detail)
            }
            peers = details
            rpc.shutdown()
        } catch {
            message = ErrorMessages.userFriendly(from: error)
        }
        isLoading = false
    }

    func addPeer(ipAddress: String, port: Int64) async {
        guard let rpc = createRpcService() else { return }
        do {
            _ = try await rpc.addPeer(ipAddress: ipAddress, port: port)
            message = "Peer added successfully"
            await loadPeers()
            rpc.shutdown()
        } catch {
            message = ErrorMessages.userFriendly(from: error)
        }
    }

    func pingPeer(peerId: Int64) async {
        guard let rpc = createRpcService() else { return }
        do {
            let success = try await rpc.pingPeer(peerId: peerId)
            message = success ? "Ping successful" : "Ping failed"
            rpc.shutdown()
        } catch {
            message = ErrorMessages.userFriendly(from: error)
        }
    }

    func activatePeer(peerId: Int64) async {
        guard let rpc = createRpcService() else { return }
        do {
            _ = try await rpc.activatePeer(peerId: peerId)
            message = "Peer activated"
            await loadPeers()
            rpc.shutdown()
        } catch {
            message = ErrorMessages.userFriendly(from: error)
        }
    }

    func deactivatePeer(peerId: Int64) async {
        guard let rpc = createRpcService() else { return }
        do {
            _ = try await rpc.deactivatePeer(peerId: peerId)
            message = "Peer deactivated"
            await loadPeers()
            rpc.shutdown()
        } catch {
            message = ErrorMessages.userFriendly(from: error)
        }
    }

    func clearMessage() { message = nil }
}

File: android/AlpineApp/app/src/main/java/com/alpine/app/ui/screens/peers/PeersViewModel.kt
package com.alpine.app.ui.screens.peers

import android.app.Application
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.alpine.app.data.model.PeerDetail
import com.alpine.app.data.rpc.AlpineRpcService
import com.alpine.app.data.rpc.TlsConfig
import com.alpine.app.data.rpc.TlsMode
import com.alpine.app.ui.screens.settings.dataStore
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

class PeersViewModel(application: Application) : AndroidViewModel(application) {

    private val dataStore = application.dataStore

    private val _peers = MutableStateFlow<List<PeerDetail>>(emptyList())
    val peers: StateFlow<List<PeerDetail>> = _peers.asStateFlow()

    private val _isLoading = MutableStateFlow(true)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _message = MutableStateFlow<String?>(null)
    val message: StateFlow<String?> = _message.asStateFlow()

    private var rpcService: AlpineRpcService? = null

    init {
        loadPeers()
    }

    private fun loadPeers() {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                val svc = getService()
                val peerIds = svc.getAllPeers()
                val details = peerIds.map { svc.getPeer(it) }
                _peers.value = details
            } catch (e: Exception) {
                _message.value = "Failed to load peers: ${e.message}"
            }
            _isLoading.value = false
        }
    }

    fun addPeer(ipAddress: String, port: Long) {
        viewModelScope.launch {
            try {
                getService().addPeer(ipAddress, port)
                _message.value = "Peer added"
                loadPeers()
            } catch (e: Exception) {
                _message.value = "Failed: ${e.message}"
            }
        }
    }

    fun pingPeer(peerId: Long) {
        viewModelScope.launch {
            try {
                val ok = getService().pingPeer(peerId)
                _message.value = if (ok) "Ping sent" else "Ping failed"
            } catch (e: Exception) {
                _message.value = "Ping error: ${e.message}"
            }
        }
    }

    fun activatePeer(peerId: Long) {
        viewModelScope.launch {
            try {
                getService().activatePeer(peerId)
                _message.value = "Peer activated"
            } catch (e: Exception) {
                _message.value = "Failed: ${e.message}"
            }
        }
    }

    fun deactivatePeer(peerId: Long) {
        viewModelScope.launch {
            try {
                getService().deactivatePeer(peerId)
                _message.value = "Peer deactivated"
            } catch (e: Exception) {
                _message.value = "Failed: ${e.message}"
            }
        }
    }

    fun clearMessage() {
        _message.value = null
    }

    private suspend fun getService(): AlpineRpcService {
        if (rpcService == null) {
            val prefs = dataStore.data.first()
            val host = prefs[stringPreferencesKey("host")] ?: "10.0.2.2"
            val port = prefs[stringPreferencesKey("port")] ?: "8080"
            val tlsEnabled = prefs[stringPreferencesKey("tls_enabled")] == "true"
            val tlsModeStr = prefs[stringPreferencesKey("tls_mode")] ?: "SYSTEM_CA"
            val certFp = prefs[stringPreferencesKey("tls_cert_fingerprint")] ?: ""
            val tlsConfig = TlsConfig(
                enabled = tlsEnabled,
                mode = try { TlsMode.valueOf(tlsModeStr) } catch (_: Exception) { TlsMode.SYSTEM_CA },
                certFingerprint = certFp,
                hostname = host
            )
            rpcService = AlpineRpcService(tlsConfig.buildUrl(host, port), tlsConfig, host)
        }
        return rpcService!!
    }

    override fun onCleared() {
        super.onCleared()
        rpcService?.shutdown()
    }
}
